doc/help/elektra-highlevel-gen.md

```diff
@@ -0,0 +1,415 @@
+ELEKTRA_GET_ARRAY_ELEMENT_SIGNATURE (ElektraEnumMyenum, EnumMyenum);
+ELEKTRA_SET_SIGNATURE (ElektraEnumMyenum, EnumMyenum);
+ELEKTRA_SET_ARRAY_ELEMENT_SIGNATURE (ElektraEnumMyenum, EnumMyenum);
+```
+
+These functions are not generated per key, but per enum type. If multiple keys use the same enum type (both need to define the full metadata,
+including the full set of values), we only generate one set of these functions.
+
+The getter and setter functions won't be explained here, they work like any of the other getter and setter functions of the high-level API.
+
+The other three functions are used to convert between the string values and the generated `enum`. You may find these useful in your
+application. You can call them via e.g. `ELEKTRA_KEY_TO (EnumMyenum)`. The difference between `ELEKTRA_TO_STRING` and
+`ELEKTRA_TO_CONST_STRING` is that the first returns a `char *` allocated via `elektraMalloc`, while the second returns a static
+`constchar *`.
+
+Both `ELEKTRA_TO_STRING` and `ELEKTRA_TO_CONST_STRING` are always implemented via a straightforward `switch` statement. The implementation
+of `ELEKTRA_KEY_TO` on the other hand can be changed via the `enumConv` parameter. If you set `enumConv=strcmp`, we will generate a code
+analogous to:
+
+```c
+if (strcmp (string, "none") == 0) { /* ... */ }
+if (strcmp (string, "red") == 0) { /* ... */ }
+if (strcmp (string, "green") == 0) { /* ... */ }
+if (strcmp (string, "blue") == 0) { /* ... */ }
+```
+
+This code is not really optimal, since we really only need to look at the first character to determine the correct enum value. This is where
+`enumConv=trie` comes in. With this option, we use a trie to generate a series of (nested, if necessary) switch/case statements:
+
+```c
+switch (string[0])
+{
+case 'b': /* blue */
+case 'g': /* green */
+case 'n': /* none */
+case 'r': /* red */
+}
+```
+
+Of course this version also has its own problems. Take for example the the enum with the values: `blue`, `blueish` and `brown`. With
+`enumConv=trie` this would generate the following code:
+
+```c
+switch (string[0])
+{
+case 'b':
+    switch (string[1])
+    {
+    case 'l':
+        switch (string[2])
+        {
+        case 'u':
+            switch (string[3])
+            {
+            case 'e':
+                switch (string[4])
+                {
+                case 'i': /* blueish */
+                }
+                /* blue */
+            }
+            break;
+        }
+        break;
+        case 'r': /* brown */
+    }
+    break;
+}
+```
+
+This is not only hard to read, but also not very efficient. Of course we could just look at `string[1]` and `string[4]` or at
+`strlen(string)` or at any number of other things. But writing a code-generator that does this for the general case is very hard. That is
+why the default option `enumConv=default` uses the trie version, if the depth is less than 3 (i.e. if looking at `string[0]` and `string[1]`
```

Comment:
No, the trie implementation is very basic. Depth `n` means looking at the first `n` characters `string[0], string[1], ..., string[n-1]`.

Reply:
Yes, then clarify in text. Maybe the compiler optimizes this anyway, so it is not so clear how efficient it is.

Btw.: I would not call this trie. Trie is a data structure and not this arrangement of switch/case statements. So better call it switch vs. strcmp. (Or leave the whole option away if you find out what is faster.)

Reply:
> Btw.: I would not call this trie. Trie is a data structure and not this arrangement of switch/case statements.

I called it trie, because internally an actual trie is used to generate the switch/case statements.

> So better call it switch vs. strcmp.

Will do.